proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,184 @@
+syscall handler will copy the entire sysvar to the given memory location.
+
+Notably, BPF programs do not inform the syscall handler of the expected size
+of the sysvar. BPF programs thus effectively assume that the requested sysvars
+will forever have the same size.
+
+This prevents future upgrades that shrink or extend sysvars. In the same
+spirit, if a new sysvar is created, a new syscall must be created.
+
+Furthermore, if certain queries are requred by BPF programs - such as
+`SlotHashes::get_slot()` and `StakeHistory::get_entry()` - those single-element
+retrievals will *also* require new syscalls.
+
+It's worth noting that at the time of this writing, the work required for
+[SIMD-0088](./0088-enable-core-bpf-programs.md) will require at least five new
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+## Detailed Design
+
+By simply redefining the `get` API to accept `offset` and `length` parameters,
+one single `get` syscall can be used by all sysvars to retrieve data.
+
+```rust
+fn get(offset: usize, length: usize) -> &[u8];
+```
+
+The specifics of returning optional values or errors can be left up to the
+implementation, but the following cases would be handled with errors:
+
+- The sysvar data is unavailable.
+- The sysavr data is corrupt.
+- The provided offset or length would be out of bounds on the sysvar data.
+
+Sysvars themselves should have internal logic for understanding their own size
+and the size of their entries if they are a list-based data structure. With this
+information, each sysvar can customize how it wishes to drive the `get` syscall.
```

Comment:
highly in favor of the unstated implication here ie that this syscall is not made available to programs directly and must be provided through an interface (though perhaps that is already the case for all syscalls?) so we dont lock ourselves into memory layout decisions more than we already are